File: Topics/HashTable/1657.py
from collections import Counter, defaultdict
class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        l1 = [0] * 26
        l2 = [0] * 26
        n1 = len(word1)
        n2 = len(word2)

        for i in range(n1):
            l1[ord("a") - ord(word1[i])] += 1
            l2[ord("a") - ord(word2[i])] += 1

        return n1 == n2 and sorted(l1) == sorted(l2) and set(word1) == set(word2)

word1 = "ab"
word2 = "aa"
slv = Solution()
print(slv.closeStrings(word1, word2))

# another solution
from collections import Counter, defaultdict
class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        l1 = Counter(word1)
        l2 = Counter(word2)
        test = l1.keys() == l2.keys()
        for key, val in l1.items():
            flag = True
            for k2, val2 in l2.items():
                if val == val2 and test:
                    flag = False
                    l2[k2] = "0"
                    break
            if flag: return False 

        return True
word1 = "cca"
word2 = "aac"
slv = Solution()
print(slv.closeStrings(word1, word2))

File: Topics/HashTable/1657.js
"use strict";
function closeStrings(word1, word2) {
    let n1 = word1.length, n2 = word2.length;
    let map1 = {}, map2 = {};
    if (n1 !== n2)
        return false;
    for (let i = 0; i < n1; i++) {
        map1[word1[i]] = (map1[word1[i]] || 0) + 1;
        map2[word2[i]] = (map2[word2[i]] || 0) + 1;
    }
    for (let key1 in map1) {
        let flag = true;
        for (let key2 in map2) {
            if (map1[key1] == map2[key2] && map2.hasOwnProperty(key1)) {
                map2[key2] = 0;
                flag = false;
                break;
            }
        }
        if (flag)
            return false;
    }
    return true;
}
;
let word1 = "aax", word2 = "ssu";
console.log(closeStrings(word1, word2));
